### PR TITLE
find mkl installed by nuget

### PR DESCRIPTION
**Pull request by cyyever**

### CODE REVIEW

File: cmake/Modules/FindMKL.cmake [hunk 1]
```diff
@@ -212,7 +216,7 @@ MACRO(CHECK_ALL_LIBRARIES LIBRARIES OPENMP_TYPE OPENMP_LIBRARY _name _list _flag
         # Separately handling compiled TBB
         SET(_found_tbb TRUE)
       ELSE()
-        FIND_LIBRARY(${_prefix}_${_library}_LIBRARY NAMES ${_library})
+        FIND_LIBRARY(${_prefix}_${_library}_LIBRARY NAMES ${_library} ${_library}_dll)
```
Comment (peterjc123) [inline cmake/Modules/FindMKL.cmake]: The DLLs with the suffix `_dll` are used for shared libraries. We use static libraries here.

Comment (cyyever) [reply cmake/Modules/FindMKL.cmake]: use BUILD_SHARED_LIBS  to check build type

File: cmake/Modules/FindMKL.cmake [hunk 2]
```diff
@@ -212,7 +216,11 @@ MACRO(CHECK_ALL_LIBRARIES LIBRARIES OPENMP_TYPE OPENMP_LIBRARY _name _list _flag
         # Separately handling compiled TBB
         SET(_found_tbb TRUE)
       ELSE()
-        FIND_LIBRARY(${_prefix}_${_library}_LIBRARY NAMES ${_library})
+        SET(lib_names ${_library})
+        IF (BUILD_SHARED_LIBS)
```
Comment (peterjc123) [inline cmake/Modules/FindMKL.cmake]: Actually we are using the static version of MKL even when we build the dynamic version(`BUILD_SHARED_LIBS=ON`). So we may need to create a new variable here.

**Comment by peterjc123** [comment]
BTW, the static version of libs are available on Nuget. We need the following two packages:
https://www.nuget.org/packages/intelmkl.static.win-x64/
https://www.nuget.org/packages/intelopenmp.devel.win/

**Review by peterjc123 — COMMENTED** [review]
Thanks for the PR. The new changes LGTM. However, I don't think it is good to link against shared MKL libs as a fallback if `BUILD_SHARED_LIBS=ON` because of the following reasons.
1. It makes other users and devs unaware of this situation.
2. The DLLs are not installed so users may face problem during initialization.

The better solutions to this problem would be
1. Remove the changes and force the linker to link against static MKL libs.
2. Add a new variable to specify runtime type (default: static) and then specify the `NAMES` of `FIND_LIBRARY` accordingly. (`${_library}` for static libs and `${_library}_dll` for shared libs, except `libiomp5md` because only the shared version is available) You may also need to install the MKL libraries when necessary.

**Comment by cyyever** [comment]
>     2\. available



> 
> 
> Thanks for the PR. The new changes LGTM. However, I don't think it is good to link against shared MKL libs as a fallback if `BUILD_SHARED_LIBS=ON` because of the following reasons.
> 
>     1. It makes other users and devs unaware of this situation.
> 
>     2. The DLLs are not installed so users may face problem during initialization.
> 
> 
> The better solutions to this problem would be
> 
>     1. Remove the changes and force the linker to link against static MKL libs.
> 
>     2. Add a new variable to specify runtime type (default: static) and then specify the `NAMES` of `FIND_LIBRARY` accordingly. (`${_library}` for static libs and `${_library}_dll` for shared libs, except `libiomp5md` because only the shared version is available) You may also need to install the MKL libraries when necessary.



> 
> 
> Thanks for the PR. The new changes LGTM. However, I don't think it is good to link against shared MKL libs as a fallback if `BUILD_SHARED_LIBS=ON` because of the following reasons.
> 
>     1. It makes other users and devs unaware of this situation.
> 
>     2. The DLLs are not installed so users may face problem during initialization.
> 
> 
> The better solutions to this problem would be
> 
>     1. Remove the changes and force the linker to link against static MKL libs.
> 
>     2. Add a new variable to specify runtime type (default: static) and then specify the `NAMES` of `FIND_LIBRARY` accordingly. (`${_library}` for static libs and `${_library}_dll` for shared libs, except `libiomp5md` because only the shared version is available) You may also need to install the MKL libraries when necessary.

I think it suffices to use the INTER_MKL_DIR variable to specify the static MKL directory,and discard the BUILD_SHARED_LIBS check, in other cases it falls back to the previous behavior. 
If you want to strictly differentiate the static and dll cases,we should need a separate variable to control this and issue a FATAL_ERROR when it finds the wrong type.

**Comment by peterjc123** [comment]
> discard the BUILD_SHARED_LIBS check

I‘m okay with discarding the changes related to `BUILD_SHARED_LIBS`.

**Comment by ezyang** [comment]
@peterjc123 if this looks good to you I'm OK with landing it. It'll be hard to prevent this from bitrotting without a test but I don't think we can conveniently test this case.

File: cmake/Modules/FindMKL.cmake [hunk 3]
```diff
@@ -212,7 +238,8 @@ MACRO(CHECK_ALL_LIBRARIES LIBRARIES OPENMP_TYPE OPENMP_LIBRARY _name _list _flag
         # Separately handling compiled TBB
         SET(_found_tbb TRUE)
       ELSE()
-        FIND_LIBRARY(${_prefix}_${_library}_LIBRARY NAMES ${_library})
+        SET(lib_names ${_library} ${_library}_dll)
+        FIND_LIBRARY(${_prefix}_${_library}_LIBRARY NAMES ${lib_names})
```
Comment (peterjc123) [inline cmake/Modules/FindMKL.cmake]: My only concern is that introducing the shared libs here may break some of our release builds. Could you please do that in a separate PR?

Comment (cyyever) [reply cmake/Modules/FindMKL.cmake]: > 
> 
> My only concern is that introducing the shared libs here may break some of our release builds. Could you please do that in a separate PR?

OK

**Review by peterjc123 — APPROVED** [review]
